### PR TITLE
Cephfs: Remove user creation for every volume

### PR DESCRIPTION
**Pull request by poornimag**
Currently, provisioner creates user for every volume and nodeplugin
uses this user to mount that volume. But nodeplugin and provisioner
already have admin credentials, hence using the admin credentials
to mount the volume and getting rid of user creation for each volume.

Signed-off-by: Poornima G <pgurusid@redhat.com>

### CODE REVIEW

**Comment by ShyamsundarR** [comment]
@poornimag the intention is to provide an option to choose between, per PV ID/key generation versus using the admin credentials for the FS to mount the volume, as detailed in issue #269 

This PR removes the entire per PV user creation which is not what was desired. Is there a different reasoning at the moment that necessitates this removal?

**Comment by poornimag** [comment]
> @poornimag the intention is to provide an option to choose between, per PV ID/key generation versus using the admin credentials for the FS to mount the volume, as detailed in issue #269
> 
> This PR removes the entire per PV user creation which is not what was desired. Is there a different reasoning at the moment that necessitates this removal?

Apologies, i wasn't aware of the approach of making user id optional. When nodeserver already has admin credentials, what is the need of creating user per volume? We create the user and use that user, so i m not sure what security issue its fixing. Also, Ramana mentioned that he would be adding subvolume group user create which would have access to a given group. Eventually we would need to use that instead of admin?

**Comment by humblec** [comment]
@ShyamsundarR I do believe per volume user creation may not be something we have to consider atm for the reasons mentioned in @poornimag's comment.

**Comment by ShyamsundarR** [comment]
> > @poornimag the intention is to provide an option to choose between, per PV ID/key generation versus using the admin credentials for the FS to mount the volume, as detailed in issue #269
> > This PR removes the entire per PV user creation which is not what was desired. Is there a different reasoning at the moment that necessitates this removal?
> 
> Apologies, i wasn't aware of the approach of making user id optional. When nodeserver already has admin credentials, what is the need of creating user per volume? We create the user and use that user, so i m not sure what security issue its fixing. Also, Ramana mentioned that he would be adding subvolume group user create which would have access to a given group. Eventually we would need to use that instead of admin?

As per my understanding when opening the issue, the intention was to make it optional to improve the performance of VolumeCreate (and Delete), it was not to do away with the per PV secret for users who needed it.

I am not aware of why there is a per PV user required or its security implications. I am tagging @liewegas @batrick and @gman0  for comments and direction.

Even if this functionality is removed for 1.1, one thing I would like to point out is that for backward compatibility reasons, 1.0 CephFS created volumes would need to use the user created for the same during mount, and so we cannot entirely delete the code.

**Comment by poornimag** [comment]
> > > @poornimag the intention is to provide an option to choose between, per PV ID/key generation versus using the admin credentials for the FS to mount the volume, as detailed in issue #269
> > > This PR removes the entire per PV user creation which is not what was desired. Is there a different reasoning at the moment that necessitates this removal?
> > 
> > 
> > Apologies, i wasn't aware of the approach of making user id optional. When nodeserver already has admin credentials, what is the need of creating user per volume? We create the user and use that user, so i m not sure what security issue its fixing. Also, Ramana mentioned that he would be adding subvolume group user create which would have access to a given group. Eventually we would need to use that instead of admin?
> 
> As per my understanding when opening the issue, the intention was to make it optional to improve the performance of VolumeCreate (and Delete), it was not to do away with the per PV secret for users who needed it.
> 
> I am not aware of why there is a per PV user required or its security implications. I am tagging @liewegas @batrick and @gman0 for comments and direction.
> 
> Even if this functionality is removed for 1.1, one thing I would like to point out is that for backward compatibility reasons, 1.0 CephFS created volumes would need to use the user created for the same during mount, and so we cannot entirely delete the code.
WRT backward compatibility, i think the admin user will be able to mount the volumes irrespective of the users that were created for the subvolume. So we may not need backward compatibility code, need to confirm.

**Comment by humblec** [comment]
Once we have subvol group user in ceph manager , we can consider adding the functionality for cephfs volumes. For now, i believe this per pv user creation or option/choice does not add much value, rather confusing.

**Comment by poornimag** [comment]
> > > > @poornimag the intention is to provide an option to choose between, per PV ID/key generation versus using the admin credentials for the FS to mount the volume, as detailed in issue #269
> > > > This PR removes the entire per PV user creation which is not what was desired. Is there a different reasoning at the moment that necessitates this removal?
> > > 
> > > 
> > > Apologies, i wasn't aware of the approach of making user id optional. When nodeserver already has admin credentials, what is the need of creating user per volume? We create the user and use that user, so i m not sure what security issue its fixing. Also, Ramana mentioned that he would be adding subvolume group user create which would have access to a given group. Eventually we would need to use that instead of admin?
> > 
> > 
> > As per my understanding when opening the issue, the intention was to make it optional to improve the performance of VolumeCreate (and Delete), it was not to do away with the per PV secret for users who needed it.
> > I am not aware of why there is a per PV user required or its security implications. I am tagging @liewegas @batrick and @gman0 for comments and direction.
> > Even if this functionality is removed for 1.1, one thing I would like to point out is that for backward compatibility reasons, 1.0 CephFS created volumes would need to use the user created for the same during mount, and so we cannot entirely delete the code.
> 
> WRT backward compatibility, i think the admin user will be able to mount the volumes irrespective of the users that were created for the subvolume. So we may not need backward compatibility code, need to confirm.

Although there is one thing to handle in backward compatibility is to delete user when we delete the old type PVC.

**Comment by humblec** [comment]
@poornimag looks like we need a rebase :)

**Comment by batrick** [comment]
> > > @poornimag the intention is to provide an option to choose between, per PV ID/key generation versus using the admin credentials for the FS to mount the volume, as detailed in issue #269
> > > This PR removes the entire per PV user creation which is not what was desired. Is there a different reasoning at the moment that necessitates this removal?
> > 
> > 
> > Apologies, i wasn't aware of the approach of making user id optional. When nodeserver already has admin credentials, what is the need of creating user per volume? We create the user and use that user, so i m not sure what security issue its fixing. Also, Ramana mentioned that he would be adding subvolume group user create which would have access to a given group. Eventually we would need to use that instead of admin?
>>
> As per my understanding when opening the issue, the intention was to make it optional to improve the performance of VolumeCreate (and Delete), it was not to do away with the per PV secret for users who needed it.

In general, CSI shouldn't be using the **admin** key anyway. That is the veritable keys-to-the-kingdom for Ceph. A CephFS credential with full access to the file system is a starting point:

http://docs.ceph.com/docs/nautilus/cephfs/client-auth/

And that's just for the nodes doing the mounts.

For the CSI driver, it could be using the admin key (because it needs to do `ceph fs volume` calls) but even ideally it should be using a watered down key. In fact, the CSI driver itself doesn't really need access to the file system at all.

@ajarr @liewegas

**Comment by ajarr** [comment]
> 
> In general, CSI shouldn't be using the **admin** key anyway. That is the veritable keys-to-the-kingdom for Ceph. A CephFS credential with full access to the file system is a starting point:
> 
> http://docs.ceph.com/docs/nautilus/cephfs/client-auth/
> 
> And that's just for the nodes doing the mounts.
> 
> For the CSI driver, it could be using the admin key (because it needs to do `ceph fs volume` calls) but even ideally it should be using a watered down key. In fact, the CSI driver itself doesn't really need access to the file system at all.
>

For `fs subvolume/volume` calls, the node would need an auth ID with 
following capabilities,
```
caps mon = "allow r"
caps mgr = "allow rw"
```

For mounting Ceph FS subvolumes within the CSI subvolume group, the node would need an auth ID with following capabilities,
```
caps mon = "allow r"
caps mds = "allow rw path=/volumes/csi"
caps osd  = "allow rw pool=cephfs.data.a"
```
were '/volumes/csi' is the backend path of the subvolume group 'csi', and 'cephfs.data.a' is the data pool of the volume/filesystem within which the subvolumes are created.


The former auth ID should be created beforehand and passed to the CephFS CSI plugin. This will allow the plugin to talk to the ceph-mgr volumes module for provisioning/managing the FS subvolumes. The CephFS CSI plugin issues the following `fs subvolumegroup` command to the ceph-mgr,
```
ceph fs subvolumegroup authorize <auth ID> <volname> <subvolname> --group_name csi 
```
to create the latter auth ID and get it's key, which will be used to mount the FS subvolumes.

**Comment by ShyamsundarR** [comment]
> > > > @poornimag the intention is to provide an option to choose between, per PV ID/key generation versus using the admin credentials for the FS to mount the volume, as detailed in issue #269
> > > > This PR removes the entire per PV user creation which is not what was desired. Is there a different reasoning at the moment that necessitates this removal?
> > > 
> > > 
> > > Apologies, i wasn't aware of the approach of making user id optional. When nodeserver already has admin credentials, what is the need of creating user per volume? We create the user and use that user, so i m not sure what security issue its fixing. Also, Ramana mentioned that he would be adding subvolume group user create which would have access to a given group. Eventually we would need to use that instead of admin?
> > 
> > 
> > As per my understanding when opening the issue, the intention was to make it optional to improve the performance of VolumeCreate (and Delete), it was not to do away with the per PV secret for users who needed it.
> 
> In general, CSI shouldn't be using the **admin** key anyway. That is the veritable keys-to-the-kingdom for Ceph. A CephFS credential with full access to the file system is a starting point:
> 
> http://docs.ceph.com/docs/nautilus/cephfs/client-auth/
> 
> And that's just for the nodes doing the mounts.
> 
> For the CSI driver, it could be using the admin key (because it needs to do `ceph fs volume` calls) but even ideally it should be using a watered down key. In fact, the CSI driver itself doesn't really need access to the file system at all.
> 
> @ajarr @liewegas

Just to be clear the final statement is, we are not supporting per subvolume/csi-volume ID creation (or rather dropping support for the same).

**Comment by liewegas** [comment]
This seems like a fine choice for CSI.  @batrick do you concur?

Of course we still need per-subvolume creds for manila, but that's a different type of environment where you don't have a [semi-]trusted intermediary doing the mounts for you.

**Comment by batrick** [comment]
> were /volumes/csi is the backend path of the subvolume group csi, and cephfs.data.a is the data pool of the volume/filesystem within which the subvolumes are created.

We can be even

> > In general, CSI shouldn't be using the **admin** key anyway. That is the veritable keys-to-the-kingdom for Ceph. A CephFS credential with full access to the file system is a starting point:
> > http://docs.ceph.com/docs/nautilus/cephfs/client-auth/
> > And that's just for the nodes doing the mounts.
> > For the CSI driver, it could be using the admin key (because it needs to do `ceph fs volume` calls) but even ideally it should be using a watered down key. In fact, the CSI driver itself doesn't really need access to the file system at all.
> 
> For `fs subvolume/volume` calls, the node would need an auth ID with
> following capabilities,
> 
> ```
> caps mon = "allow r"
> caps mgr = "allow rw"
> ```
> 
> For mounting Ceph FS subvolumes within the CSI subvolume group, the node would need an auth ID with following capabilities,
> 
> ```
> caps mon = "allow r"
> caps mds = "allow rw path=/volumes/csi"
> caps osd = "allow rw pool=cephfs.data.a"
> ```
> 
> were `/volumes/csi` is the backend path of the subvolume group `csi`, and `cephfs.data.a` is the data pool of the volume/filesystem within which the subvolumes are created.

Keep in mind that the CSI provisioner may be creating volumes in different pools to achieve different QoS/redundancy. I think this should just be:

`caps osd = "allow rw tag cephfs data=cephfs"`

So that the nodes can access any data pool used by the file system using the single credential.

> The former auth ID should be created beforehand and passed to the CephFS CSI plugin. This will allow the plugin to talk to the ceph-mgr volumes module for provisioning/managing the FS subvolumes. The CephFS CSI plugin issues the following `fs subvolumegroup` command to the ceph-mgr,
> 
> ```
> ceph fs subvolumegroup authorize <auth ID> <volname> <subvolname> --group_name csi 
> ```
> 
> to create the latter auth ID and get its key, which will be used to mount the FS subvolumes.

As long as the subvolume module is issuing an `fs authorize` in the backend, that should setup things properly.

**Comment by batrick** [comment]
> This seems like a fine choice for CSI. @batrick do you concur?
> 
> Of course we still need per-subvolume creds for manila, but that's a different type of environment where you don't have a [semi-]trusted intermediary doing the mounts for you.

Right. CSI doesn't need multiple credentials. It just adds complexity. 

Manila definitely will though but only for kernel/ceph-fuse mounts. Eventually when we complete work on the nfs clustering automagic setup, the ganesha daemons will have a pre-configured credential setup by the ceph-mgr. Manila can be blissfully uninvolved.

@tombarron

**Comment by ajarr** [comment]
> This will allow the plugin to talk to the ceph-mgr volumes module for provisioning/managing the FS subvolumes. The CephFS CSI plugin issues the following `fs subvolumegroup` command to the ceph-mgr,
> 
> ```
> ceph fs subvolumegroup authorize <auth ID> <volname> <subvolname> --group_name csi 
> ```
> 
> to create the latter auth ID and get its key, which will be used to mount the FS subvolumes.

@ShyamsundarR , I've implemented this command. The [PR](https://github.com/ceph/ceph/pull/28973) is in review. Would CSI use case require evicting clients that have mounted using the common auth ID?

**Comment by ShyamsundarR** [comment]
> > This will allow the plugin to talk to the ceph-mgr volumes module for provisioning/managing the FS subvolumes. The CephFS CSI plugin issues the following `fs subvolumegroup` command to the ceph-mgr,
> > ```
> > ceph fs subvolumegroup authorize <auth ID> <volname> <subvolname> --group_name csi 
> > ```
> > 
> > 
> > to create the latter auth ID and get its key, which will be used to mount the FS subvolumes.
> 
> @ShyamsundarR , I've implemented this command. The [PR](https://github.com/ceph/ceph/pull/28973) is in review. Would CSI use case require evicting clients that have mounted using the common auth ID?

Do we need this? As the code would **not** be invoking this directly, as we are not creating per subvolume IDs and keys.

Also, the user credentials that are needed to invoke CephFS operations by the CSI plugin, are passed in te required secrets. I would assume generating these credentials is an admin operation, prior to passing the id/key into the secrets and that as long as we document how to create an ID with required permissions (as in #360 ), we should be fine.

What am I missing here?

**Comment by ajarr** [comment]
> > > This will allow the plugin to talk to the ceph-mgr volumes module for provisioning/managing the FS subvolumes. The CephFS CSI plugin issues the following `fs subvolumegroup` command to the ceph-mgr,
> > > ```
> > > ceph fs subvolumegroup authorize <auth ID> <volname> <subvolname> --group_name csi 
> > > ```
> > > 
> > > 
> > > to create the latter auth ID and get its key, which will be used to mount the FS subvolumes.
> > 
> > 
> > @ShyamsundarR , I've implemented this command. The [PR](https://github.com/ceph/ceph/pull/28973) is in review. Would CSI use case require evicting clients that have mounted using the common auth ID?
> 
> Do we need this? As the code would **not** be invoking this directly, as we are not creating per subvolume IDs and keys.
> 
> Also, the user credentials that are needed to invoke CephFS operations by the CSI plugin, are passed in te required secrets. I would assume generating these credentials is an admin operation, prior to passing the id/key into the secrets and that as long as we document how to create an ID with required permissions (as in #360 ), we should be fine.
> 
> What am I missing here?

Agree. The admin (or operator) can create one auth ID with capabilities to issue ceph-mgr fs subvolume commands, and to mount the fs subvolumes. The auth IDs capabilites would be,
```
caps mon = "allow r"
caps mgr = "allow rw"
caps mds = "allow rw path=/volumes/csi"
caps osd = "allow rw tag cephfs data=cephfs "
```
assuming the ceph volume/filesystem name is cephfs.

@poornimag , you've no dependencies on ceph-mgr development. You should be able to move this PR forward.

@ShyamsundarR , do we expect that users would create snapshots of files/directories within the PVs? If so then, the OSD caps needs to be "allow rwp" instead of "allow rw". Then again if users create snapshots within the PV, the PV deletion would fail if snapshots exist within it.

**Comment by ShyamsundarR** [comment]
> @poornimag , you've no dependencies on ceph-mgr development. You should be able to move this PR forward.

@poornimag are we handling backward compat in this patch or is that going to be a separate patch? As discussed earlier, use the created IDs to mount or use the provided credentials to mount, but on deletes the legacy volume IDs need to be cleaned up.

> 
> @ShyamsundarR , do we expect that users would create snapshots of files/directories within the PVs? If so then, the OSD caps needs to be "allow rwp" instead of "allow rw". Then again if users create snapshots within the PV, the PV deletion would fail if snapshots exist within it.

No, the users just get to see an FS that they can operate on, they are not meant to create snapshots of objects etc. within the same. If at all in the future, if the need to do so arises, modifying the caps should be enough to get this working (I never see this happening, but just noting it nevertheless).

**Comment by ajarr** [comment]
> > @ShyamsundarR , do we expect that users would create snapshots of files/directories within the PVs? If so then, the OSD caps needs to be "allow rwp" instead of "allow rw". Then again if users create snapshots within the PV, the PV deletion would fail if snapshots exist within it.
> 
> No, the users just get to see an FS that they can operate on, they are not meant to create snapshots of objects etc. within the same. If at all in the future, if the need to do so arises, modifying the caps should be enough to get this working (I never see this happening, but just noting it nevertheless).

@ShyamsundarR , OK. To clarify, it's the MDS caps that need an additional 'p' flag to set layout and quota, and 's' flag to create snapshots.
http://docs.ceph.com/docs/master/cephfs/client-auth/#layout-and-quota-restriction-the-p-flag

**Comment by humblec** [comment]
> @poornimag , you've no dependencies on ceph-mgr development. You should be able to move this PR forward.

Based on above comment, @poornimag can you please revisit this PR ?

**Comment by poornimag** [comment]
@ShyamsundarR  Backward compatibility is taken care of.

As per the discussion, we would use admin credentials for create/mount/delete of pvc instead of per subvolumegroup user provided by ceph-mgr.

**Comment by ajarr** [comment]
> @ShyamsundarR Backward compatibility is taken care of.
> 
> As per the discussion, we would use admin credentials for create/mount/delete of pvc instead of per subvolumegroup user provided by ceph-mgr.

@poornimag , what discussion? can you please elaborate the reason here to stick to admin credentials?

**Comment by ajarr** [comment]
@batrick I was referring to this comment and its replies,
https://github.com/ceph/ceph-csi/pull/437#issuecomment-511202488

**Comment by ajarr** [comment]
> > @ShyamsundarR Backward compatibility is taken care of.
> > As per the discussion, we would use admin credentials for create/mount/delete of pvc instead of per subvolumegroup user provided by ceph-mgr.
> 
> @poornimag , what discussion? can you please elaborate here on the reason to stick to admin credentials?

OK. Shyam clarified. The admin credentials refers to credentials passed to Ceph-CSI, and not ceph's admin key credentials.

File: pkg/cephfs/controllerserver.go [hunk 1]
```diff
@@ -188,12 +182,12 @@ func (cs *ControllerServer) deleteVolumeDeprecated(req *csi.DeleteVolumeRequest)
...
 	}
 
-	if err = deleteCephUser(&ce.VolOptions, cr, volID); err != nil {
-		klog.Errorf("failed to delete ceph user for volume %s: %v", volID, err)
+	if err = cs.MetadataStore.Delete(string(volID)); err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	if err = cs.MetadataStore.Delete(string(volID)); err != nil {
+	if err = deleteCephUserDeprecated(&ce.VolOptions, cr, volID); err != nil {
```
Comment (Madhu-1) [inline pkg/cephfs/controllerserver.go]: @poornimag do `cs.MetadataStore.Delete` after userdelete

**Comment by ajarr** [comment]
Great! Thanks @poornimag 

And now @ShyamsundarR we'll get to see how good CephFS CSI is ;-)

**Pull request merged**